src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -0,0 +1,912 @@
+    }
+
+    #[cfg(not(test))]
+    // This is the main loop of the `Exited` state.
+    fn exited(&mut self) -> StateMachine<Self> {
+        // Wait indefinitely.
+        // The VMM thread will kill the entire process.
+        let barrier = Barrier::new(2);
+        barrier.wait();
+
+        StateMachine::finish()
+    }
+
+    #[cfg(test)]
+    // In tests the main/vmm thread exits without 'exit()'ing the whole process.
+    // All channels get closed on the other side while this Vcpu thread is still running.
+    // This Vcpu thread should just do a clean finish without reporting back to the main thread.
+    fn exit(&mut self, _: u8) -> StateMachine<Self> {
+        self.exit_evt.write(1).unwrap();
+        // State machine reached its end.
+        StateMachine::finish()
+    }
+
+    /// Runs the vCPU in KVM context and handles the kvm exit reason.
+    ///
+    /// Returns error or enum specifying whether emulation was handled or interrupted.
+    pub fn run_emulation(&self) -> Result<VcpuEmulation> {
+        match self.kvm_vcpu.fd.run() {
+            Ok(run) => match run {
+                VcpuExit::MmioRead(addr, data) => {
+                    if let Some(mmio_bus) = &self.kvm_vcpu.mmio_bus {
+                        mmio_bus.read(addr, data);
+                        METRICS.vcpu.exit_mmio_read.inc();
+                    }
+                    Ok(VcpuEmulation::Handled)
+                }
+                VcpuExit::MmioWrite(addr, data) => {
+                    if let Some(mmio_bus) = &self.kvm_vcpu.mmio_bus {
+                        mmio_bus.write(addr, data);
+                        METRICS.vcpu.exit_mmio_write.inc();
+                    }
+                    Ok(VcpuEmulation::Handled)
+                }
+                VcpuExit::Hlt => {
+                    info!("Received KVM_EXIT_HLT signal");
+                    Ok(VcpuEmulation::Stopped)
+                }
+                VcpuExit::Shutdown => {
+                    info!("Received KVM_EXIT_SHUTDOWN signal");
+                    Ok(VcpuEmulation::Stopped)
+                }
+                // Documentation specifies that below kvm exits are considered
+                // errors.
+                VcpuExit::FailEntry => {
+                    // Hardware entry failure.
+                    METRICS.vcpu.failures.inc();
+                    error!("Received KVM_EXIT_FAIL_ENTRY signal");
+                    Err(Error::FaultyKvmExit(format!("{:?}", VcpuExit::FailEntry)))
+                }
+                VcpuExit::InternalError => {
+                    // Failure from the Linux KVM subsystem rather than from the hardware.
+                    METRICS.vcpu.failures.inc();
+                    error!("Received KVM_EXIT_INTERNAL_ERROR signal");
+                    Err(Error::FaultyKvmExit(format!(
+                        "{:?}",
+                        VcpuExit::InternalError
+                    )))
+                }
+                r => {
```

Comment:
nit: Maybe rename `r` to `uncommon_reason` or `arch_specific_reason`.

Reply:
Renamed.